dbt/adapters/snowflake/impl.py

```diff
@@ -82,6 +85,13 @@ def _get_warehouse(self) -> str:
 
     def _use_warehouse(self, warehouse: str):
         """Use the given warehouse. Quotes are never applied."""
+        kwargs = {
+            'warehouse': warehouse
+        }
+        warehouse = self.execute_macro(
+            SNOWFLAKE_WAREHOUSE_MACRO_NAME,
+            kwargs=kwargs
+        )
         self.execute('use warehouse {}'.format(warehouse))
```

Comment:
Rather than a macro that returns the name of the warehouse, and then templating + executing `use warehouse ...` right from within the Python context, I think it would be better to have the Jinja macro template out the actual SQL we want to run (`use warehouse {{ warehouse }}`) _and_ execute that SQL, via `statement` or `run_query`

That way, `self.execute_macro` accomplishes the full intended functionality, and we don't need the implicit call to `self.execute` (a.k.a. `adapter.execute`).

The macro could be named `use_warehouse` or `use_snowflake_warehouse`. To accomplish your intended functionality, you could still override that macro, adding complex conditional rules to determine the warehouse name, as long as the macro still templates out `use warehouse ...` and executes that SQL.

Reply:
@javiCalvo Sorry for the delayed response on my end!

Fair point. It is in closer keeping with how other adapter methods/macros work in dbt today, but I also appreciate wanting a narrower macro that _just_ templates the desired SQL.

Another way we could factor this:
```sql
{% macro get_use_warehouse_sql(warehouse) -%}
    use warehouse {{ warehouse }}
{%- endmacro %}
```
```python
    def _use_warehouse(self, warehouse: str):
        """Use the given warehouse. Quotes are never applied."""
        kwargs = {
            'warehouse': warehouse
        }
        use_warehouse_sql = self.execute_macro(
            SNOWFLAKE_USE_WAREHOUSE_MACRO_NAME,
            kwargs=kwargs
        )
        self.execute(use_warehouse_sql)
```
Then your override can be just this (returning a string, no `call`, much easier to test):
```sql
{% macro get_use_warehouse_sql(warehouse) -%}
   {% set warehouse = warehouse | replace('_X_', '_' + {'stg': 'D', 'tst': 'Q', 'prd': 'P'}[target.name] + '_') %}
    use warehouse {{ warehouse }}
{%- endmacro %}
```